zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1101,7 +1104,10 @@ protected void processStateChanged(WatchedEvent event) {
         fireAllEvents();
       }
     } else if (event.getState() == KeeperState.Expired) {
-      reconnectOnExpiring();
+      _isNewSessionEventFired = false;
```

Comment:
What if reconnect is not done successfully in the following reconnectOnExpiring() call?
In that case, if we change this flag to false, will it cause additional fireNewSessionEvents()? For example if we have a stale new session event in the queue.

Reply:
Discussed wit Kai offline, this should be fine as long as processStateChanged is synchronized called in one thread.

Reply:
BTW, please update the comment of _isNewSessionEventFired, it is out of date.

Reply:
I was also thinking about the same concern.  It should be fine as it is protected by the eventLock. And also if reconnecting fails because of IOException, the old/expired zk is kept, but the expired zk could not do anything.

Reply:
updated.